addons/hw_drivers/iot_handlers/blackbox/data_reader.py

```diff
@@ -0,0 +1,184 @@
+import threading
+
+from serial.threaded import Protocol
+from .helpers import log
+
+
+STX = 0x02
+ETX = 0x03
+ACK = 0x06
+NACK = 0x15
+T1_MAX = 0.3
+T2_MAX = 0.75
+RETRY_MAX = 3
+
+
+class DataReader(Protocol):
+    def __init__(self):
+        super().__init__()
+        self.buffer = bytearray()
+
+        # key -> {"event": threading.Event, "response": str | None}
+        self._on_stop_callbacks = []
+        self._pending = {}
+        self._pending_lock = threading.Lock()
+
+    def connection_lost(self, exc):
+        log('error', 'DataReader', f'Connection lost with {self.transport.serial.port}')
+        for callback in self._on_stop_callbacks:
+            callback()
+        super().connection_lost(exc)
```

Comment:
Looks like super().connection_lost(exc) isn't needed as per documentation `serial.threaded.Protocol` class provides empty implementations of all methods.

`exc` can be used to get more information about the reason of the connection loss (of not `None`)